E5 - Carteirinha

AdrielSchneider opened the pull request:
Desenvolvido processo de carteirinha.

Falta apenas a criação dos testes unitários e gostaríamos de implementar um validação para que a carteirinha possa ser vinculada a um usuário válido.

Além do mais, atualmente, o processo está aceitando o idUsuario como Int, devido ao fato de faltarem as classes disponibilizadas pelas outras equipes (Acredito que a coluna que o campo que iremos utilizar será o id da classe Pessoa, disponibilizado pela Equipe 3.

Desenvolvido em conjunto com @kevin120

AdrielSchneider committed:
Adicionado processo de Carteirinha

AdrielSchneider committed:
Implementado processo de carteirinha

Implementado processo de carteirinha

AdrielSchneider committed:
Implementada validação de formato do id

Implementada validação de formato do id

AdrielSchneider committed:
Ajustados imports e removido código não utilizado

Ajustados imports e removido código não utilizado.

AdrielSchneider committed:
Melhorias nos padrões e implementação de validações

Melhorias nos padrões e implementação de validações

kevin120 committed:
Implementado processo para registro de extravio de carteirinha

AdrielSchneider committed:
Adicionadas anotações

Adicionadas anotações nos métodos referentes á Carrteirinha.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/model/Carteirinha.kt:
```diff
@@ -0,0 +1,62 @@
+
+    fun registrarEntrega(): Carteirinha {
+
+        if (status != StatusCarteirinha.ENTREGA_PENDENTE)
+            throw Exception("Não é possível entregar uma carteirinha que não esteja pendente")
+
+        if (Date() <= dataEmissao)
+            throw Exception("Não é possível registrar a carteirinha com data de entrega menor que data de emissão")
```
Retornar outra classe de exception. Ex: IllegalStateException

AdrielSchneider replied:
Ajustado tipo da exceção retornada.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/model/Carteirinha.kt:
```diff
@@ -0,0 +1,62 @@
+        return copy(status = StatusCarteirinha.VALIDA, dataEntrega = Date());
+    }
+
+    fun registrarExtravio(): Carteirinha {
+        return copy(status = StatusCarteirinha.EXTRAVIADA);
+    }
```
É para ser possível registrar um extravio de uma carteirinha com status "Pendente de entrega"?

AdrielSchneider replied:
Chegamos a considerar uma validação para esse ponto, mas não a aplicamos devido a não ser um critério de aceite, o que nos levou a pensar na possibilidade da pessoa que gerencia a emissão/entrega das carteirinhas poder acabar perdendo alguma, tendo assim de registrar a perda da mesma.

Acredito que o @djonathanb poderia esclarecer a nossa dúvida, deve ser possível registrar ou não o extravio de uma carteirinha que ainda não foi entregue?

djonathanb replied:
Faltou mesmo, não deve permitir, excelente trabalho pessoal

AdrielSchneider replied:
Ajustado para que seja possível registrar o extravio somente de carteirinhas Válidas.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/model/CarteirinhaRepository.kt:
```diff
@@ -0,0 +1,12 @@
+@Repository
+interface CarteirinhaRepository : JpaRepository<Carteirinha, String> {
+    @Query("select c from Carteirinha c where c.idUsuario = ?1 and (c.status = 'VALIDA' or c.status = 'ENTREGA_PENDENTE')")
+    abstract fun findByIdUsuario(idUsuario: Int): Optional<Carteirinha>
```
Talvez trocar o nome do método para explicitar a query.
Ex: findCarteirinhaValidaByIdUSuario,
Ou até receber os status por parâmetro, e usar o nome do método como findByIdUsuarioAndStatusIn, ai você nem mesmo precisa colocar a anotação @query

AdrielSchneider replied:
Alterado nome do método.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/model/MotivoExtravio.kt:
```diff
@@ -0,0 +1,25 @@
+        @Column(nullable = false)
+        val motivoExtravio: String) {
+
+    fun registrarExtravio(): MotivoExtravio {
+        val m = MotivoExtravio(dataExtravio = Date(), numeroCarteirinha = numeroCarteirinha, motivoExtravio = motivoExtravio);
+
+        return m
+    }
```
Essa função não faz exatamente o que o construtor faz?

AdrielSchneider replied:
Alterado para que, no lugar do método, seja usado somente o o construtor do objeto MotivoExtravio.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/services/impl/CarteirinhaQueryServiceImpl.kt:
```diff
@@ -0,0 +1,24 @@
+
+    override fun findByIdUsuario(idUsuario: Int): Carteirinha {
+        return carteirinhaRepository.findByIdUsuario(idUsuario).orElseThrow{
+            Exception("Nenhuma carteirinha Válida encontrada para o Usuário $idUsuario")
```
Utilizar a classe NotFoundException que o professor criou

AdrielSchneider replied:
Criada nova classe de exceção para que seja disparada ao não encontrar carteirinha para o usuário informado.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/services/impl/CarteirinhaApplicationServiceImpl.kt:
```diff
@@ -0,0 +1,98 @@
+        return date.time
+    }
+
+    fun validaRegex(value: String): Boolean {
+        if (!Pattern.compile("[A-Z]{2}[0-9]{7}[-][0-9]{2}").matcher(value).matches())
+            throw Exception("Número de carteirinha não segue padrão determinado! Ex: UF9999999-99")
+
+        return true;
```
Utilizar a anotação @Pattern do bean validation, ai pode remover esse trecho aqui.
https://cheatsheetseries.owasp.org/cheatsheets/Bean_Validation_Cheat_Sheet.html

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/services/impl/CarteirinhaQueryServiceImpl.kt:
```diff
@@ -0,0 +1,24 @@
+@Service
+class CarteirinhaQueryServiceImpl(private val carteirinhaRepository: CarteirinhaRepository): CarteirinhaQueryService {
+    override fun buscaPorId(idCarteirinha: String): Carteirinha = carteirinhaRepository.findById(idCarteirinha).orElseThrow {
+        Exception("Carteirinha com o id ${idCarteirinha} não encontrada!")
```
Utilizar a classe NotFoundException

AdrielSchneider replied:
Criada nova classe de exceção para o caso.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/RegistraExtravioUseCase.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Service
+interface RegistraExtravioUseCase {
+    fun RegistraExtravio(carteirinha: Carteirinha): Carteirinha
```
colocar inicial do método minúsculo.

AdrielSchneider replied:
Corrigido nome do método para que a primeira letra seja minúscula.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/RegistraMotivoExtravioUseCase.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Service
+interface RegistraMotivoExtravioUseCase {
+    fun RegistraMotivoExtravio(motivoExtravio: MotivoExtravio): MotivoExtravio
```
colocar inicial do método minúsculo.

AdrielSchneider replied:
Corrigido nome do método para que a primeira letra seja minúscula.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/CriaCarteirinhaUseCaseImpl.kt:
```diff
@@ -0,0 +1,24 @@
+        var carteirinhaResult = carteirinhaRepository.findByIdUsuario(carteirinha.idUsuario)
+
+        if (carteirinhaResult.isPresent)
+            throw Exception("Usuário ${carteirinha.idUsuario} já possui uma carteirinha válida: ${carteirinhaResult.get().numeroCarteirinha}")
```
Utilizar outra classe de exceção, ou ainda, criar uma para esse caso

AdrielSchneider replied:
Criada classe de exceção.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/CriaCarteirinhaUseCaseImpl.kt:
```diff
@@ -0,0 +1,24 @@
+        carteirinhaResult = carteirinhaRepository.findById(carteirinha.numeroCarteirinha)
+
+        if (carteirinhaResult.isPresent)
+            throw Exception("A carteirinha ${carteirinha.numeroCarteirinha} já está cadastrada")
```
o número da carteirinha é uma Primary Key, não vai aceitar valores duplicados, é realmente necessário manter esse trecho no código?

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/RegistraCarteirinhaUseCaseImpl.kt:
```diff
@@ -0,0 +1,25 @@
+        var carteirinhaResult: Carteirinha
+        try {
+            carteirinhaResult = carteirinhaQueryService.buscaPorId(carteirinha.numeroCarteirinha)
+        } catch (ex: Exception) {
+            throw ex;
+        }
```
Recolhe a exceção e depois joga de volta?

AdrielSchneider replied:
Removido.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/ValidaCarteirinhaUseCaseImpl.kt:
```diff
@@ -0,0 +1,26 @@
+        val carteirinhaResult: Carteirinha
+        try {
+            carteirinhaResult = carteirinhaQueryService.buscaPorId(carteirinha.numeroCarteirinha)
+        } catch (ex: Exception) {
+            throw ex;
+        }
```
ajustar esse trecho também

AdrielSchneider replied:
Removido.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/ValidaCarteirinhaUseCaseImpl.kt:
```diff
@@ -0,0 +1,26 @@
+            throw ex;
+        }
+
+        if (!carteirinhaResult.validate())
+            throw Exception("Carteirinha Inválida:\n Status: ${carteirinhaResult.status}")
```
Se precisa retornar uma exceção, faça o método validate() jogar uma exceção

AdrielSchneider replied:
Corrigido para que a exceção seja disparada de dentro do método validate()

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/interfaces/rest/carteirinha/CarteirinhaResource.kt:
```diff
@@ -0,0 +1,74 @@
+class CarteirinhaResource(private val carteirinhaApplicationService: CarteirinhaApplicationService) {
+
+    @Operation(summary = "Cria uma nova carteirinha para o usuário e retorna o endereço do novo recurso")
+    @ApiResponses(value = [
+        ApiResponse(responseCode = "201", description = "Carteirinha criada com sucesso!")
+    ])
+    @PostMapping
+    fun criarCarteirinha(@RequestBody @Valid dto: CarteirinhaDTO): ResponseEntity<Any> {
+
+        try {
+
+            val idNovaCarteirinha = carteirinhaApplicationService.criar(dto)
+
+            val uri = ServletUriComponentsBuilder.fromCurrentRequest()
+                    .path("/{id}")
+                    .buildAndExpand(idNovaCarteirinha)
+                    .toUri()
+
+            return ResponseEntity.created(uri).build()
+
+        } catch (error: Exception) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error.message)
+        }
+
+    }
```
Se o método pode retornar NotFound, então falta a documentação desse retorno

AdrielSchneider replied:
Documentados retornos possíveis.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/interfaces/rest/carteirinha/CarteirinhaResource.kt:
```diff
@@ -0,0 +1,74 @@
+
+    }
+
+    @Operation(summary = "Verifica se uma carteirinha é válida ou não")
+    @PostMapping("/validate")
+    fun verificarValidade(@RequestBody @Valid dto: CarteirinhaDTO) : ResponseEntity<Any> {
+        try {
+            val carteirinha = carteirinhaApplicationService.validaCarteirinha(dto)
+
+            return ResponseEntity.ok(carteirinha)
+        } catch (ex: Exception) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ex.message)
+        }
+    }
```
Falta a documentação dos retornos HTTP

AdrielSchneider replied:
Documentados retornos possíveis.

BrunoSzczuk commented on src/main/kotlin/br/unipar/plano/interfaces/rest/carteirinha/CarteirinhaResource.kt:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    @Operation(summary = "Registra a entrega da carteirinha ao usuário")
+    @PostMapping("/register")
+    fun registrarEntrega(@RequestBody @Valid dto: CarteirinhaDTO): ResponseEntity<Any> {
+        try {
+            val carteirinha = carteirinhaApplicationService.registraEntrega(dto)
+            return ResponseEntity.ok(carteirinha)
+        } catch (ex: Exception) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ex.message)
+        }
+    }
+
+    @Operation(summary = "Registra o extravio e grava o motivo deste")
+    @PostMapping("/registermisplacement")
+    fun registrarExtravio(@RequestBody @Valid motivoDTO: MotivoDTO): ResponseEntity<Any> {
+        try {
+            val motivoExtravio = carteirinhaApplicationService.registraExtravio(motivoDTO)
+            return ResponseEntity.ok(motivoExtravio)
+        } catch (ex: Exception) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ex.message)
+        }
+    }
```
Falta a documentação dos retornos HTTP

AdrielSchneider replied:
Documentados retornos possíveis.

BrunoSzczuk reviewed (changes requested):
Parabéns pelo desenvolvimento!
Deixei alguns pontos de melhoria.

AdrielSchneider committed:
Corrigido nome do método

Corrigido nome do método

djonathanb commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/RegistraExtravioUseCaseImpl.kt:
```diff
@@ -0,0 +1,17 @@
+
+    override fun registraExtravio(carteirinha: Carteirinha): Carteirinha {
+
+        var carteirinhaResult: Carteirinha = carteirinha.registrarExtravio()
```
Você pode receber somente o id e consultar ela aqui no caso de uso

djonathanb commented on src/main/kotlin/br/unipar/plano/domain/carteirinha/usecases/impl/ValidaCarteirinhaUseCaseImpl.kt:
```diff
@@ -0,0 +1,26 @@
+    override fun validar(carteirinha: Carteirinha): Carteirinha {
+        val carteirinhaResult: Carteirinha
+        try {
+            carteirinhaResult = carteirinhaQueryService.buscaPorId(carteirinha.numeroCarteirinha)
```
Pode receber somente o númeroc como parâmetro

djonathanb commented on src/main/kotlin/br/unipar/plano/interfaces/rest/carteirinha/CarteirinhaResource.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Operation(summary = "Verifica se uma carteirinha é válida ou não")
+    @PostMapping("/validate")
```
Ao invés de Post você pode fazer um get que recebe como parâmetro "/{numeroCarteirinha}/valida" e parar de passar o DTO pelo body.

djonathanb commented on src/main/kotlin/br/unipar/plano/interfaces/rest/carteirinha/CarteirinhaResource.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Operation(summary = "Registra a entrega da carteirinha ao usuário")
+    @PostMapping("/register")
```
Não coloque verbos na url, melhor @PostMappint("/{numeroCarteirinha}/entrega"), não me parece que você precisa de algo do corpo da requisição.

AdrielSchneider committed:
Realizadas correções propostas pelo professor

- Utilizado tipo de variável LocalDate.
- Corrigidas funções e métodos para que recebam somente o número da carteirinha como parâmetro, facilitando assim o uso dos métodos/endpoints.

AdrielSchneider committed:
Implementadas classes de teste.

Implementadas classes de teste das carteirinhas.

kevin120 committed:
Ajustado para que só seja possível registrar extravio de uma carteirinha válida

Ajustado para que só seja possível registrar extravio de uma carteirinha válida

kevin120 committed:
Ajustado testes para o registro de extravio da carteirinha

Ajustado testes para o registro de extravio da carteirinha

kevin120 committed:
Feito alguns ajustes solicitados no code review

Feito alguns ajustes solicitados no code review

AdrielSchneider committed:
Corrigido título de teste

Corrigido título de teste

AdrielSchneider committed:
Adicionada anotação dos verbos Http

Adicionada anotação dos verbos Http

AdrielSchneider committed:
Alterado nome do método

Alterado nome do método

AdrielSchneider committed:
Alterado caminho das classes de exceção criadas

Alterado caminho das classes de exceção criadas

AdrielSchneider committed:
Criadas exceções

Criadas exceções

AdrielSchneider committed:
Removidas tratativas desnecessárias

Removidas tratativas desnecessárias
